controllers/vsphereclusteridentity_controller.go

```diff
@@ -0,0 +1,179 @@
+	var (
+		controllerNameShort = fmt.Sprintf("%s-controller", strings.ToLower(identityControlledTypeName))
+		controllerNameLong  = fmt.Sprintf("%s/%s/%s", ctx.Namespace, ctx.Name, controllerNameShort)
+	)
+
+	// Build the controller context.
+	controllerContext := &context.ControllerContext{
+		ControllerManagerContext: ctx,
+		Name:                     controllerNameShort,
+		Recorder:                 record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+		Logger:                   ctx.Logger.WithName(controllerNameShort),
+	}
+
+	reconciler := clusterIdentityReconciler{ControllerContext: controllerContext}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(identityControlledType).
+		WithOptions(controller.Options{MaxConcurrentReconciles: ctx.MaxConcurrentReconciles}).
+		Complete(reconciler)
+}
+
+type clusterIdentityReconciler struct {
+	*context.ControllerContext
+}
+
+func (r clusterIdentityReconciler) Reconcile(ctx _context.Context, req reconcile.Request) (_ reconcile.Result, reterr error) {
+	// Get VSphereClusterIdentity
+	identity := &infrav1.VSphereClusterIdentity{}
+	if err := r.Client.Get(r, req.NamespacedName, identity); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Logger.V(4).Info("VSphereClusterIdentity not found, won't reconcile", "key", req.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Create the patch helper.
+	patchHelper, err := patch.NewHelper(identity, r.Client)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(
+			err,
+			"failed to init patch helper for %s %s/%s",
+			identity.GroupVersionKind(),
+			identity.Namespace,
+			identity.Name)
+	}
+
+	defer func() {
+		conditions.SetSummary(identity, conditions.WithConditions(infrav1.CredentialsAvailableCondidtion))
+
+		if err := patchHelper.Patch(ctx, identity); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
+			r.Logger.Error(err, "patch failed", "namespace", identity.Namespace, "name", identity.Name)
+		}
+	}()
+
+	if !identity.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, identity)
+	}
+
+	// fetch secret
+	secret := &corev1.Secret{}
+	secretKey := client.ObjectKey{
+		Namespace: r.Namespace,
+		Name:      identity.Spec.SecretName,
+	}
+	if err := r.Client.Get(ctx, secretKey, secret); err != nil {
+		conditions.MarkFalse(identity, infrav1.CredentialsAvailableCondidtion, infrav1.SecretNotAvailableReason, clusterv1.ConditionSeverityWarning, err.Error())
+	}
+
+	if !clusterutilv1.IsOwnedByObject(secret, identity) {
+		if len(secret.OwnerReferences) > 0 {
+			conditions.MarkFalse(identity, infrav1.CredentialsAvailableCondidtion, infrav1.SecretAlreadyInUseReason, clusterv1.ConditionSeverityError, "secret being used by another Cluster/VSphereIdentity")
+			identity.Status.Ready = false
+			return reconcile.Result{}, errors.New("secret being used by another Cluster/VSphereIdentity")
+		}
+
+		identitySecretMU.Lock()
```

Comment:
do we need a lock here ? I wouldn't expect concurrent reconciles of the same `VSphereClusterIdentity` to happen at the same time in the controller. If it's to deal with other identities adding the owner refs to this secret that should be handled by optimistic concurrency at the apiserver level + your check above

Reply:
It was to deal with other identities adding themselves as the owner. I've removed it. An identical flow exists with clusters that use secrets directly. That's also currently utilizing a lock. Should that also be removed?

Reply:
yes